File: rust_tm_api/benchmarks/src/bank.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;
use tm::{TmCell, Transaction, transaction, tm_init, tm_exit};

// ── Configuration ───────────────────────────────────────
const DEFAULT_DURATION_MS: u64 = 10000;
const DEFAULT_NB_ACCOUNTS: usize = 1024;
const DEFAULT_NB_THREADS: usize = 4;
const DEFAULT_READ_ALL: u32 = 20;
const DEFAULT_INITIAL_BALANCE: i32 = 1000;

// ── Account & Bank ──────────────────────────────────────
struct Account {
    _number: i32,
    balance: TmCell<i32>,
}

impl Account {
    fn new(number: i32) -> Self {
        Account {
            _number: number,
            balance: TmCell::new(DEFAULT_INITIAL_BALANCE),
        }
    }
}

struct Bank {
    accounts: Vec<Account>,
}

impl Bank {
    fn new(size: usize) -> Self {
        let mut accounts = Vec::with_capacity(size);
        for i in 0..size {
            accounts.push(Account::new(i as i32));
        }
        Bank { accounts }
    }

    fn size(&self) -> usize {
        self.accounts.len()
    }
}

// ── TX operations ───────────────────────────────────────
fn transfer(tx: &Transaction, bank: &Bank, src: usize, dst: usize, amount: i32) {
    let b = tx.read(&bank.accounts[src].balance);
    tx.write(&bank.accounts[src].balance, b - amount);

    let b = tx.read(&bank.accounts[dst].balance);
    tx.write(&bank.accounts[dst].balance, b + amount);
}

fn total_transactional(tx: &Transaction, bank: &Bank) -> i32 {
    let mut total = 0i32;
    for account in &bank.accounts {
        total += tx.read(&account.balance);
    }
    total
}

fn total_non_transactional(bank: &Bank) -> i32 {
    // Direct access without TM — only called outside any transaction
    let mut total = 0i32;
    // SAFETY: We read via raw pointer to bypass TM (peek equivalent)
    for account in &bank.accounts {
        unsafe {
            total += *account.balance.ptr();
        }
    }
    total
}

// ── Main ────────────────────────────────────────────────
fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut duration_ms = DEFAULT_DURATION_MS;
    let mut nb_accounts = DEFAULT_NB_ACCOUNTS;
    let mut nb_threads = DEFAULT_NB_THREADS;
    let mut read_all_pct = DEFAULT_READ_ALL;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "-d" if i + 1 < args.len() => {
                i += 1;
                duration_ms = args[i].parse().unwrap_or(DEFAULT_DURATION_MS);
            }
            "-a" if i + 1 < args.len() => {
                i += 1;
                nb_accounts = args[i].parse().unwrap_or(DEFAULT_NB_ACCOUNTS);
            }
            "-t" if i + 1 < args.len() => {
                i += 1;
                nb_threads = args[i].parse().unwrap_or(DEFAULT_NB_THREADS);
            }
            "-r" if i + 1 < args.len() => {
                i += 1;
                read_all_pct = args[i].parse().unwrap_or(DEFAULT_READ_ALL);
            }
            "-h" | "--help" => {
                println!("Usage: bank [-d ms] [-a n] [-t n] [-r pct]");
                return;
            }
            _ => {}
        }
        i += 1;
    }

    println!("Bank Benchmark — Rust TM API");
    println!(
        "Duration: {} ms  Accounts: {}  Threads: {}",
        duration_ms, nb_accounts, nb_threads
    );

    tm_init();

    let bank = Arc::new(Bank::new(nb_accounts));
    let expected_total = (nb_accounts as i32) * DEFAULT_INITIAL_BALANCE;
    let initial_total = total_non_transactional(&bank);
    println!("Initial total: {}  Expected: {}", initial_total, expected_total);
    assert_eq!(initial_total, expected_total, "Initial total mismatch");

    let stop = Arc::new(AtomicBool::new(false));

    let mut handles = Vec::with_capacity(nb_threads);
    for tid in 0..nb_threads {
        let bank = Arc::clone(&bank);
        let stop = Arc::clone(&stop);
        handles.push(std::thread::spawn(move || {
            let mut rng = fastrand::Rng::new();
            rng.seed(tid as u64 + 1234);
            let n = bank.size();
            let mut txfer = 0u64;
            let mut reads = 0u64;

            while !stop.load(Ordering::Relaxed) {
                let roll = rng.f64() * 100.0;
                if roll < read_all_pct as f64 {
                    transaction(|tx| {
                        total_transactional(tx, &bank);
                    });
                    reads += 1;
                } else {
                    let src = rng.usize(0..n);
                    let mut dst = rng.usize(0..n);
                    if dst == src {
                        dst = (src + 1) % n;
                    }
                    transaction(|tx| {
                        transfer(tx, &bank, src, dst, 1);
                    });
                    txfer += 1;
                }
            }
            (txfer, reads)
        }));
    }

    std::thread::sleep(Duration::from_millis(duration_ms));
    stop.store(true, Ordering::Relaxed);

    let mut total_txfer = 0u64;
    let mut total_reads = 0u64;
    for h in handles {
        let (t, r) = h.join().unwrap();
        total_txfer += t;
        total_reads += r;
    }

    let final_total = total_non_transactional(&bank);
    let total_txns = total_txfer + total_reads;

    let aborts = tm::tm_abort_count();
    println!(
        "\nResults:\n  Final total: {}  Expected: {}\n  Transfers: {}  Read-all: {}\n  Total txns: {}\n  TM aborts: {}",
        final_total, expected_total, total_txfer, total_reads, total_txns, aborts
    );

    tm_exit();

    if final_total == expected_total {
        println!("PASS: Money conserved");
    } else {
        eprintln!(
            "FAIL: Money {} by {}",
            if final_total > expected_total { "created" } else { "destroyed" },
            if final_total > expected_total {
                final_total - expected_total
            } else {
                expected_total - final_total
            }
        );
        std::process::exit(1);
    }
}

File: expli_benchmarks/tests/bank/bank.cpp
#include "expli_tm_api/tm_api.hpp"
#include "expli_tm_api/tx_executor.hpp"
#include <atomic>
#include <chrono>
#include <cstdio>
#include <cstring>
#include <functional>
#include <condition_variable>
#include <mutex>
#include <random>
#include <thread>

// ── Configuration ───────────────────────────────────────────
constexpr int DEFAULT_DURATION_MS  = 10000;
constexpr int DEFAULT_NB_ACCOUNTS  = 1024;
constexpr int DEFAULT_NB_THREADS   = 4;
constexpr int DEFAULT_READ_ALL     = 20;
constexpr int DEFAULT_WRITE_ALL    = 0;
constexpr int DEFAULT_INITIAL_BALANCE = 1000;

// ── Account & Bank ──────────────────────────────────────────
struct Account {
    int number;
    expli::TM<int> balance;
};

struct Bank {
    expli::vector<Account> accounts;

    Bank(int size) {
        accounts.resize(size);
        for (int i = 0; i < size; ++i) {
            accounts[i].number = i;
            accounts[i].balance.poke(DEFAULT_INITIAL_BALANCE);
        }
    }
    int size() const { return (int)accounts.size(); }
};

Bank *g_bank;   // set once in main, never modified

// ── Executor (optional, set by --queue) ─────────────────────
static expli::TxExecutor *g_executor = nullptr;

// Dispatch to either the inline TM::transaction or the executor
template<typename F>
void run_tx(F&& body) {
    if (g_executor) {
        tx_transaction(*g_executor, std::forward<F>(body));
    } else {
        expli::TM<int>::transaction(std::forward<F>(body));
    }
}

// ── TX functions ────────────────────────────────────────────
void transfer(int src, int dst, int amount) {
    run_tx([&]() {
        Account *a_src = &g_bank->accounts[src];
        Account *a_dst = &g_bank->accounts[dst];

        int bal = a_src->balance.read();
        bal -= amount;
        a_src->balance.write(bal);

        bal = a_dst->balance.read();
        bal += amount;
        a_dst->balance.write(bal);
    });
}

int total_transactional() {
    int total = 0;
    run_tx([&]() {
        total = 0;
        for (int i = 0; i < g_bank->size(); ++i)
            total += g_bank->accounts[i].balance.read();
    });
    return total;
}

void reset() {
    run_tx([&]() {
        for (int i = 0; i < g_bank->size(); ++i)
            g_bank->accounts[i].balance.write(DEFAULT_INITIAL_BALANCE);
    });
}

int total_non_transactional() {
    // Read directly without TM (only called outside any TX)
    int total = 0;
    for (int i = 0; i < g_bank->size(); ++i)
        total += g_bank->accounts[i].balance.peek();
    return total;
}

// ── Barrier ─────────────────────────────────────────────────
class Barrier {
    std::mutex mtx_;
    std::condition_variable cv_;
    int count_;
    int crossing_{0};
public:
    explicit Barrier(int n) : count_(n) {}
    void wait() {
        std::unique_lock<std::mutex> lock(mtx_);
        if (++crossing_ < count_) {
            cv_.wait(lock);
        } else {
            crossing_ = 0;
            cv_.notify_all();
        }
    }
};

// ── Control ─────────────────────────────────────────────────
std::atomic<bool> g_stop{false};

// ── Thread Data ─────────────────────────────────────────────
struct ThreadData {
    Barrier *barrier;
    std::atomic<uint64_t> nb_transfer;
    std::atomic<uint64_t> nb_read_all;
    std::atomic<uint64_t> nb_write_all;
    unsigned int seed;
    int thread_id, read_all_pct, write_all_pct, nb_threads;
    bool disjoint;

    ThreadData(Barrier *b, unsigned s, int tid, int ra, int wa, int nt, bool dis)
        : barrier(b), nb_transfer(0), nb_read_all(0), nb_write_all(0),
          seed(s), thread_id(tid), read_all_pct(ra), write_all_pct(wa),
          nb_threads(nt), disjoint(dis) {}
};

// ── Worker ──────────────────────────────────────────────────
void worker_thread(ThreadData &data) {
    expli::TM<int>::thread_init();

    auto rng = std::mt19937(data.seed);

    int rand_max = data.disjoint ? (g_bank->size() / data.nb_threads) : g_bank->size();
    int rand_min = data.disjoint ? (rand_max * data.thread_id) : 0;

    data.barrier->wait();

    while (!g_stop.load()) {
        double roll = std::uniform_real_distribution<double>(0, 100.0)(rng);
        if (roll < data.read_all_pct) {
            total_transactional();
            data.nb_read_all.fetch_add(1);
        } else if (roll < data.read_all_pct + data.write_all_pct) {
            reset();
            data.nb_write_all.fetch_add(1);
        } else {
            std::uniform_int_distribution<> ad(0, rand_max - 1);
            int src = ad(rng) + rand_min;
            int dst = ad(rng) + rand_min;
            if (dst == src)
                dst = ((src + 1) % rand_max) + rand_min;
            transfer(src, dst, 1);
            data.nb_transfer.fetch_add(1);
        }
    }

    expli::TM<int>::thread_exit();
}

// ── Main ────────────────────────────────────────────────────
int main(int argc, char *argv[]) {
    int duration_ms = DEFAULT_DURATION_MS;
    int nb_accounts = DEFAULT_NB_ACCOUNTS;
    int nb_threads  = DEFAULT_NB_THREADS;
    int read_all_pct  = DEFAULT_READ_ALL;
    int write_all_pct = DEFAULT_WRITE_ALL;
    bool disjoint = false;
    bool queue_mode = false;

    for (int i = 1; i < argc; ++i) {
        if      (!strcmp(argv[i], "-d") && i+1 < argc) duration_ms  = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-a") && i+1 < argc) nb_accounts  = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-t") && i+1 < argc) nb_threads   = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-r") && i+1 < argc) read_all_pct = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-w") && i+1 < argc) write_all_pct = atoi(argv[++i]);
        else if (!strcmp(argv[i], "--disjoint")) disjoint = true;
        else if (!strcmp(argv[i], "--queue")) queue_mode = true;
        else if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help")) {
            printf("Usage: bank [-d ms] [-a n] [-t n] [-r pct] [-w pct] [--disjoint] [--queue]\n");
            return 0;
        }
    }

    printf("Bank Benchmark — Explicit TM API\n");
    printf("Duration: %d ms  Accounts: %d  Threads: %d\n", duration_ms, nb_accounts, nb_threads);

    if (nb_accounts < nb_threads && disjoint) {
        fprintf(stderr, "Error: accounts < threads for disjoint mode\n");
        return 1;
    }

    // Queue executor (if --queue)
    expli::QueueExecutor *qexec = nullptr;
    if (queue_mode) {
        qexec = new expli::QueueExecutor(nb_threads, nb_threads);
        g_executor = qexec;
        printf("Mode: queue (%d workers, %d queues)\n", nb_threads, nb_threads);
    } else {
        printf("Mode: inline\n");
    }

    expli::TM<int>::init();

    g_bank = new Bank(nb_accounts);
    Barrier barrier(nb_threads);

    int expected_total = nb_accounts * DEFAULT_INITIAL_BALANCE;
    int initial_total  = total_non_transactional();
    printf("Initial total: %d  Expected: %d\n", initial_total, expected_total);
    if (initial_total != expected_total) { fprintf(stderr, "ERROR\n"); return 1; }

    void *td_mem = operator new[](nb_threads * sizeof(ThreadData));
    ThreadData *td = (ThreadData*)td_mem;
    for (int i = 0; i < nb_threads; ++i)
        new (&td[i]) ThreadData(&barrier, (unsigned)(i+1234), i, read_all_pct, write_all_pct, nb_threads, disjoint);

    auto start = std::chrono::high_resolution_clock::now();
    void *thr_mem = operator new[](nb_threads * sizeof(std::thread));
    std::thread *threads = (std::thread*)thr_mem;
    for (int i = 0; i < nb_threads; ++i)
        new (&threads[i]) std::thread(worker_thread, std::ref(td[i]));

    std::this_thread::sleep_for(std::chrono::milliseconds(duration_ms));
    g_stop.store(true);
    for (int i = 0; i < nb_threads; ++i) threads[i].join();

    int final_total = total_non_transactional();
    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();

    uint64_t txfer=0, reads=0, writes=0;
    for (int i = 0; i < nb_threads; ++i) {
        txfer += td[i].nb_transfer.load();
        reads += td[i].nb_read_all.load();
        writes+= td[i].nb_write_all.load();
    }

    for (int i = 0; i < nb_threads; ++i) {
        td[i].~ThreadData();
        threads[i].~thread();
    }
    operator delete[](td_mem);
    operator delete[](thr_mem);
    uint64_t total_txns = txfer + reads + writes;

    printf("\nResults:\n");
    printf("  Elapsed: %lld ms\n", (long long)elapsed);
    printf("  Final total: %d  Expected: %d\n", final_total, expected_total);
    printf("  Transfers: %llu  Read-all: %llu  Write-all: %llu\n",
           (unsigned long long)txfer, (unsigned long long)reads, (unsigned long long)writes);
    printf("  Total txns: %llu  Txns/sec: %.0f\n",
           (unsigned long long)total_txns, total_txns*1000.0/elapsed);

    delete g_bank;
    if (qexec) {
        g_executor = nullptr;
        delete qexec;
    }
    expli::TM<int>::exit();

    if (final_total == expected_total) {
        printf("PASS: Money conserved\n");
        return 0;
    }
    fprintf(stderr, "FAIL: Money %s by %d\n",
            final_total > expected_total ? "created" : "destroyed",
            final_total > expected_total ? final_total - expected_total : expected_total - final_total);
    return 1;
}
